ray-operator/controllers/ray/raycluster_controller.go

```diff
@@ -204,59 +221,32 @@ func (r *RayClusterReconciler) rayClusterReconcile(ctx context.Context, request
 				"DeletionTimestamp", instance.ObjectMeta.DeletionTimestamp)
 
 			// Delete the head Pod if it exists.
-			headPods := corev1.PodList{}
-			filterLabels := client.MatchingLabels{utils.RayClusterLabelKey: instance.Name, utils.RayNodeTypeLabelKey: string(rayv1.HeadNode)}
-			if err := r.List(ctx, &headPods, client.InNamespace(instance.Namespace), filterLabels); err != nil {
+			deletedHeads, err := r.DeleteAllPods(ctx, instance.Namespace, client.MatchingLabels{
+				utils.RayClusterLabelKey:  instance.Name,
+				utils.RayNodeTypeLabelKey: string(rayv1.HeadNode),
+			})
+			if err != nil {
 				return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
 			}
-
-			for _, headPod := range headPods.Items {
-				if !headPod.DeletionTimestamp.IsZero() {
-					r.Log.Info(fmt.Sprintf("The head Pod %s is already being deleted. Skip deleting this Pod.", headPod.Name))
-					continue
-				}
-				r.Log.Info(fmt.Sprintf(
-					"Delete the head Pod %s before the Redis cleanup. "+
-						"The storage namespace %s in Redis cannot be fully deleted if the GCS process on the head Pod is still writing to it.",
-					headPod.Name, headPod.Annotations[utils.RayExternalStorageNSAnnotationKey]))
-				if err := r.Delete(ctx, &headPod); err != nil {
-					return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
-				}
-			}
-
 			// Delete all worker Pods if they exist.
-			for _, workerGroup := range instance.Spec.WorkerGroupSpecs {
-				workerPods := corev1.PodList{}
-				filterLabels = client.MatchingLabels{utils.RayClusterLabelKey: instance.Name, utils.RayNodeGroupLabelKey: workerGroup.GroupName}
-				if err := r.List(ctx, &workerPods, client.InNamespace(instance.Namespace), filterLabels); err != nil {
-					return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
-				}
-
-				for _, workerPod := range workerPods.Items {
-					if !workerPod.DeletionTimestamp.IsZero() {
-						r.Log.Info(fmt.Sprintf("The worker Pod %s is already being deleted. Skip deleting this Pod.", workerPod.Name))
-						continue
-					}
-					r.Log.Info(fmt.Sprintf(
-						"Delete the worker Pod %s. This step isn't necessary for initiating the Redis cleanup process.", workerPod.Name))
-					if err := r.Delete(ctx, &workerPod); err != nil {
-						return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
-					}
-				}
+			_, err = r.DeleteAllPods(ctx, instance.Namespace, client.MatchingLabels{
+				utils.RayClusterLabelKey:  instance.Name,
+				utils.RayNodeTypeLabelKey: string(rayv1.WorkerNode),
+			})
+			if err != nil {
+				return ctrl.Result{RequeueAfter: DefaultRequeueDuration}, err
 			}
-
-			// If the number of head Pods is not 0, wait for it to be terminated before initiating the Redis cleanup process.
-			if len(headPods.Items) != 0 {
+			if deletedHeads > 0 {
 				r.Log.Info(fmt.Sprintf(
-					"Wait for the head Pod %s to be terminated before initiating the Redis cleanup process. "+
```

Comment:
Let's keep the log message the same. A RayCluster should have at most 1 head Pod. If a RayCluster has more than 1 head Pod, it must be an unexpected behavior of KubeRay caused by a race condition. In addition, the information of the Pod name is more useful for troubleshooting.

Reply:
Restored.